sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java

```diff
@@ -0,0 +1,195 @@
+ * the License.
+ */
+
+package com.google.cloud.dataflow.sdk.runners.inprocess;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.cloud.dataflow.sdk.Pipeline;
+import com.google.cloud.dataflow.sdk.runners.PipelineRunner;
+import com.google.cloud.dataflow.sdk.runners.inprocess.InProcessPipelineRunner.CommittedBundle;
+import com.google.cloud.dataflow.sdk.runners.inprocess.InProcessPipelineRunner.UncommittedBundle;
+import com.google.cloud.dataflow.sdk.testing.TestStream;
+import com.google.cloud.dataflow.sdk.testing.TestStream.ElementEvent;
+import com.google.cloud.dataflow.sdk.testing.TestStream.Event;
+import com.google.cloud.dataflow.sdk.testing.TestStream.EventType;
+import com.google.cloud.dataflow.sdk.testing.TestStream.ProcessingTimeEvent;
+import com.google.cloud.dataflow.sdk.testing.TestStream.WatermarkEvent;
+import com.google.cloud.dataflow.sdk.transforms.AppliedPTransform;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.windowing.BoundedWindow;
+import com.google.cloud.dataflow.sdk.util.WindowedValue;
+import com.google.cloud.dataflow.sdk.util.WindowingStrategy;
+import com.google.cloud.dataflow.sdk.values.PBegin;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.dataflow.sdk.values.PCollection.IsBounded;
+import com.google.cloud.dataflow.sdk.values.PInput;
+import com.google.cloud.dataflow.sdk.values.POutput;
+import com.google.cloud.dataflow.sdk.values.TimestampedValue;
+import com.google.common.base.Supplier;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * The {@link TransformEvaluatorFactory} for the {@link TestStream} primitive.
+ */
+class TestStreamEvaluatorFactory implements TransformEvaluatorFactory {
+  private final AtomicBoolean inUse = new AtomicBoolean(false);
+  private final AtomicReference<Evaluator<?>> evaluator = new AtomicReference<>();
+
+  @Nullable
+  @Override
+  public <InputT> TransformEvaluator<InputT> forApplication(
+      AppliedPTransform<?, ?, ?> application,
+      @Nullable CommittedBundle<?> inputBundle,
+      InProcessEvaluationContext evaluationContext) throws Exception {
+    return createEvaluator((AppliedPTransform) application, evaluationContext);
+  }
+
+  private <InputT, OutputT> TransformEvaluator<? super InputT> createEvaluator(
+      AppliedPTransform<PBegin, PCollection<OutputT>, TestStream<OutputT>> application,
+      InProcessEvaluationContext evaluationContext) {
+    if (evaluator.get() == null) {
+      Evaluator<OutputT> createdEvaluator = new Evaluator<>(application, evaluationContext, inUse);
+      evaluator.compareAndSet(null, createdEvaluator);
+    }
+    if (inUse.compareAndSet(false, true)) {
```

Comment:
This logic is somewhat confusing -- can you add javadoc on why you need to do this? At face value, creating an evaluator should be as simple as "return new Evaluator<>(application, evaluationContext, inUse)". I then see that you're caching the evaluator. That makes sense. But I'm not sure why you return null if you've already created one.

(1) Just make which ever thread successfully creates the evaluator be the one to return it?
(2) Or just have a boolean, and only the thread that flips false->true returns the evaluator.

(And also comment why you want to return after the first one)